Fix prompts

corrideat opened the pull request:
Closes #2768

corrideat committed:
Fix prompts

Closes #2768

taoeffect commented on shared/domains/chelonia/utils.js:
```diff
@@ -575,8 +575,10 @@ export function eventsAfter (contractID: string, sinceHeight: number, limit?: nu
       const msg = `${eventsResponse.status}: ${eventsResponse.statusText}`
-      if (eventsResponse.status === 410) throw new ChelErrorResourceGone(msg)
-      throw new ChelErrorUnexpectedHttpResponseCode(msg)
+      // $FlowFixMe[extra-arg]
+      if (eventsResponse.status === 404 || eventsResponse.status === 410) throw new ChelErrorResourceGone(msg, { cause: eventsResponse.status })
+      // $FlowFixMe[extra-arg]
+      throw new ChelErrorUnexpectedHttpResponseCode(msg, { cause: eventsResponse.status })
```
Note: these changes need to be ported to `libcheloniajs`

taoeffect reviewed (changes requested):
Nice work @corrideat! I had a chance to test this and found some issues:

### Problem 1: supposed to be logged out?

So, I tested this on FFDev:

1. Start server with `NODE_ENV=production GI_PERSIST=sqlite grunt dev`
2. Create group, post some messages to chatroom
3. Stop server, rename `groupincome.db` to `groupincome-back.db`
4. Start server, refresh page

I got this prompt with a "Continue" button:

> Your account seems to have been deleted from the server.
> To fix this, you need to log out and create a new account.

I clicked "Continue" but was immediately logged out:

![Screenshot 2025-05-19 at 6 51 29 PM](https://github.com/user-attachments/assets/53bf5c06-a948-40ad-81af-cd8fffa93456)

If this intentional, the message needs to be changed and instead of "Continue" it should say "Log out".

If it's unintentional, the bug should be fixed.

### Problem 2: 

I stopped the server and renamed `groupincome-back.db` back to `groupincome.db`. Then I attempted to log back in with the account `u1`, however I couldn't and was greeted with this error message:

![Screenshot 2025-05-19 at 6 52 24 PM](https://github.com/user-attachments/assets/a0530df8-ac06-40e7-9786-118dd19a1fc2)

Seems like I'm permanently stuck being logged out and can't log back in even though the data is back. This should be fixed too as we don't want users permanently stuck if the sysadmin had to wipe the db and restore from backup.

corrideat committed:
Feedback

corrideat commented:
> Problem 1: supposed to be logged out?

Thanks for pointing this out. Yes, you're supposed to be logged out, and I've adjusted the text accordingly.

Before, you would get a prompt asking you if you'd like to log out, with 'Yes' and 'No'. (Except that it didn't work, because 'Prompt' wasn't allowed in the pre-login state, leading to the blank page in the issue). After our discussion, I thought that we could keep the prompt, but there was no point (for now at least) in the 'No' option, so removed it and changed 'Yes' to 'Continue'. I concur that 'Log out' is a much better label than 'Continue'.

> Problem 2

Interesting. Thanks for the steps. This was quite reproducible, and I also a bad issue. I've fixed it now by making a better check for deciding whether to call 'retain' or 'release'. This fixes this issue and in general is a more robust check should the state get corrupted.

Working on this issue showed me that logging in is in some cases rough around the edges (for example, after a deletion / restore, as you found out). Along with the improvements in #2700 / #2709, I think we should test edge cases better and improve the login process.

corrideat committed:
Merge branch 'master' into 2768-failure-to-recover-from-server-wipe

taoeffect commented on frontend/controller/actions/identity.js:
```diff
@@ -406,7 +406,7 @@ export default (sbp('sbp/selectors/register', {
 
       try {
-        if (!state) {
+        if (!cheloniaState?.contracts[identityContractID]?.references) {
```
It was at least somewhat followable before, but now this line seems to come out of nowhere.

Could you please add a detailed comment explaining why we're checking this instead of `state`?

corrideat replied:
This is what fixes the second bug. I'll add a comment explaining.

corrideat replied:
Would this comment look OK?

```
// When syncing from scratch, we need to call `retain` on our own identity
// contract to ensure the reference count is positive and we don't accidentally
// remove it.
// When not syncing from scratch, we call `sync` instead, since we don't want to
// increment the reference count every time the app is loaded.
// To decide what to call, we check if the reference count exists. This check
// is correct and also ensures that, if the state gets corrupted for some reason
// (for example, because the contract is deleted from the server and then
// restored) and the reference count lost, we call `/retain` again.
```

taoeffect replied:
Yes, I would just add, in addition to that, an explanation for why checking `!state` is insufficient.

EDIT: I realize that the comment in your head might explain that, but what I'm looking for is something like, "We don't simply check `!state` because _____"

taoeffect commented on frontend/controller/app/identity.js:
```diff
@@ -515,7 +518,15 @@ export default (sbp('sbp/selectors/register', {
+          Object.keys(cheloniaState.contracts || {}).forEach((contractID) => {
+            delete state[contractID]
+          })
+          delete state.contracts
```
Is there a bug this is preventing? If so, could you update the comment to mention it?

corrideat replied:
Not necessarily preventing a bug, but this is unnecessary information (see in `loadState` that it's discarded) and because of that, it's also potentially confusing.

taoeffect reviewed (changes requested):
Fantastic! Works well in my tests.

Just two RFCs and it looks good to merge.

corrideat committed:
Feedback

taoeffect reviewed (approved):
Fantastic!!
